spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/storage/GcpStorageAutoConfiguration.java

```diff
@@ -67,16 +73,39 @@ public GcpStorageAutoConfiguration(
         gcpStorageProperties.getCredentials().hasKey()
             ? new DefaultCredentialsProvider(gcpStorageProperties)
             : credentialsProvider;
+
+    this.universeDomain = gcpStorageProperties.getUniverseDomain();
+    this.host = gcpStorageProperties.getHost();
   }
 
   @Bean
   @ConditionalOnMissingBean
   public Storage storage() throws IOException {
-    return StorageOptions.newBuilder()
-        .setHeaderProvider(new UserAgentHeaderProvider(GcpStorageAutoConfiguration.class))
-        .setProjectId(this.gcpProjectIdProvider.getProjectId())
-        .setCredentials(this.credentialsProvider.getCredentials())
-        .build()
-        .getService();
+    StorageOptions.Builder storageOptionsBuilder =
+        StorageOptions.newBuilder()
+            .setHeaderProvider(new UserAgentHeaderProvider(GcpStorageAutoConfiguration.class))
+            .setProjectId(this.gcpProjectIdProvider.getProjectId())
+            .setCredentials(this.credentialsProvider.getCredentials());
+
+    if (this.universeDomain != null) {
+      storageOptionsBuilder.setUniverseDomain(this.universeDomain);
+    }
+    if (this.host != null) {
+      storageOptionsBuilder.setHost(verifyAndFetchHost(this.host));
+    }
+    return storageOptionsBuilder.build().getService();
+  }
+
+  private String verifyAndFetchHost(String host) {
+    URL url;
+    try {
+      url = new URL(host);
+    } catch (MalformedURLException e) {
+      throw new IllegalArgumentException(
+          "Invalid host format: "
+              + host
+              + ". Please verify that the specified host follows the 'https://${service}.${universeDomain}' format");
+    }
+    return url.getProtocol() + "://" + url.getHost() + "/";
```

Comment:
Did we get this format from Storage client library code?

Reply:
I know we have it in [java-core](https://github.com/googleapis/sdk-platform-java/blob/main/java-core/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java#L837-L851), is it where get it from?

Reply:
Ah thanks for the question. From doing a check with breakpoints in a spring storage app, Storage calls `getResolvedApiaryHost()` when setting the host at https://github.com/googleapis/java-storage/blob/7d47307d10e285f92b3e2125c4993b6a292001bf/google-cloud-storage/src/main/java/com/google/cloud/storage/HttpStorageOptions.java#L267. 

`getResolvedApiaryHost()` returns a [host with a `/` ](https://github.com/googleapis/sdk-platform-java/blob/9ad9591a63ee07449bc587b6bc15ced05dcc23a4/java-core/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java#L874) so applied the same behavior here.

Reply:
Thanks! Do you adding some comment regarding where we get the format from? I think a link to the implementation of `getResolvedApiaryHost()` in java-core should be good enough. 
In addition, anywhere we mention the format of `host`, we should always have an ending slash, which I see that it is missing in the Javadoc of [universeDomain](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/3287/files#diff-825dbfac6d19ff9cba714227e49181b02a36de47010352a4cb8f0098fbce91a3R43).

Reply:
Yup! Added a javadoc and also modified other locations where we mention host follow a consistent style.